### PR TITLE
add SDIFF and SDIFFSTORE.

### PR DESCRIPTION
**Pull request by tisilent**
relation #178

### CODE REVIEW

File: libs/server/Storage/Session/ObjectStore/SetOps.cs [hunk 1]
```diff
@@ -429,5 +432,57 @@ public GarnetStatus SetMembers<TObjectContext>(byte[] key, ArgSlice input, ref G
...
+
+        public GarnetStatus SetDiff<TObjectContext>(ArgSlice[] keys,out HashSet<byte[]> output, ref TObjectContext objectContext)
+            where TObjectContext : ITsavoriteContext<byte[], IGarnetObject, SpanByte, GarnetObjectStoreOutput, long>
+        {
+            output = default;
+            if (keys.Length == 0)
+                return GarnetStatus.OK;
+
+            output = [];
+            var comparer = new BytesComparer();
```
Comment (PaulusParssinen) [inline libs/server/Storage/Session/ObjectStore/SetOps.cs:444]: I believe you can use [`ByteArrayComparer`](https://github.com/microsoft/garnet/blob/05643c36e636dd6914653e408aa27a9c73d3ddf4/libs/server/Resp/ByteArrayComparer.cs#L13) here.

Comment (tisilent) [reply libs/server/Storage/Session/ObjectStore/SetOps.cs:450]: wow

File: libs/server/Resp/Objects/SetCommands.cs [hunk 2]
```diff
@@ -437,5 +437,123 @@ private unsafe bool SetPop<TGarnetApi>(int count, byte* ptr, ref TGarnetApi stor
...
+            }
+
+            if (NetworkKeyArraySlotVerify(ref keys, true))
+            {
+                var bufSpan = new ReadOnlySpan<byte>(recvBufferPtr, bytesRead);
+                if (!DrainCommands(bufSpan, count)) return false;
+                return true;
+            }
+
+            var status = storageApi.SetDiff(keys, out var output);
```
Comment (TalZaccai) [inline libs/server/Resp/Objects/SetCommands.cs:471]: This call needs to be fixed to match the updated method signature in IGarnetApi, as well as prep the input object (see reference in other methods in this class)

File: libs/server/Resp/Objects/SetCommands.cs [hunk 3]
```diff
@@ -437,5 +437,123 @@ private unsafe bool SetPop<TGarnetApi>(int count, byte* ptr, ref TGarnetApi stor
...
+
+            if (status == GarnetStatus.OK)
+            {
+                if (output == null || output.Count == 0)
```
Comment (TalZaccai) [inline libs/server/Resp/Objects/SetCommands.cs:475]: Here the output should be processed differently once the IGarnetApi method is updated (see reference in other methods in this class)

File: libs/server/Resp/Objects/SetCommands.cs [hunk 4]
```diff
@@ -437,5 +437,123 @@ private unsafe bool SetPop<TGarnetApi>(int count, byte* ptr, ref TGarnetApi stor
...
+        /// <typeparam name="TGarnetApi"></typeparam>
+        /// <param name="count"></param>
+        /// <param name="ptr"></param>
+        /// <param name="storageApi"></param>
+        /// <returns></returns>
+        private unsafe bool SetDiff<TGarnetApi>(int count, byte* ptr, ref TGarnetApi storageApi)
+            where TGarnetApi : IGarnetApi
+        {
+            if (count < 1)
+            {
```
Comment (TalZaccai) [inline libs/server/Resp/Objects/SetCommands.cs:452]: Reset session counters here: setItemsDoneCount = setOpsCount = 0;

File: libs/server/Resp/Objects/SetCommands.cs [hunk 5]
```diff
@@ -437,5 +437,123 @@ private unsafe bool SetPop<TGarnetApi>(int count, byte* ptr, ref TGarnetApi stor
...
+                    }
+                }
+            }
+            else if (status == GarnetStatus.WRONGTYPE)
+            {
+                // todo use WriteError?
+                while (!RespWriteUtils.WriteResponse(CmdStrings.RESP_WRONG_TYPE, ref dcurr, dend))
+                    SendAndReset();
+            }
+
```
Comment (TalZaccai) [inline libs/server/Resp/Objects/SetCommands.cs:497]: Reset session counters here as well: setItemsDoneCount = setOpsCount = 0;

File: libs/server/Resp/Objects/SetCommands.cs [hunk 6]
```diff
@@ -437,5 +437,123 @@ private unsafe bool SetPop<TGarnetApi>(int count, byte* ptr, ref TGarnetApi stor
...
+            // Move input head
+            readHead = (int)(ptr - recvBufferPtr);
+
+            return true;
+        }
+
+        private unsafe bool SetDiffStore<TGarnetApi>(int count, byte* ptr, ref TGarnetApi storageApi)
```
Comment (TalZaccai) [inline libs/server/Resp/Objects/SetCommands.cs:504]: Similar comments as in SetDiff...

File: libs/server/Storage/Session/ObjectStore/SetOps.cs [hunk 7]
```diff
@@ -429,5 +432,150 @@ public GarnetStatus SetMembers<TObjectContext>(byte[] key, ArgSlice input, ref G
...
+            finally
+            {
+                if (createTransaction)
+                    txnManager.Commit(true);
+            }
+
+            return GarnetStatus.OK;
+        }
+
+        private (HashSet<byte[]>, GarnetStatus) _setDiff<TObjectContext>(ArgSlice[] keys, ref TObjectContext objectContext)
```
Comment (TalZaccai) [inline libs/server/Storage/Session/ObjectStore/SetOps.cs:542]: This is not how we interact with Garnet objects... You need to call the StorageSession API and implement the object-specific logic in SetObjectImpl.cs. Please review other methods in this class for reference.

**Review by TalZaccai — CHANGES_REQUESTED** [review]
Thank your for your contribution! I've added a few comments for your to address. Please resubmit a request for review once you're done (also, please address the build issues...)

**Comment by tisilent** [comment]
Thank ,many doubts have answers nom.

File: libs/server/API/IGarnetApi.cs [hunk 8]
```diff
@@ -1177,6 +1186,13 @@ public interface IGarnetReadApi
         /// <returns></returns>
         GarnetStatus SetScan(ArgSlice key, long cursor, string match, int count, out ArgSlice[] items);
 
+        /// <summary>
+        /// Returns the members of the set resulting from the difference between the first set and all the successive sets.
+        /// </summary>
+        /// <param name="keys"></param>
+        /// <param name="members"></param>
+        /// <returns></returns>
+        GarnetStatus SetDiff(ArgSlice[] keys, out HashSet<byte[]> members);
```
[the remark this answers is not in the document]
Comment (tisilent) [reply libs/server/API/IGarnetApi.cs:1195]: reference SortedSetDifference..............

[the remark this answers is not in the document]
Comment (tisilent) [reply libs/server/API/IGarnetApi.cs:1195]: GarnetStatus SetDiff(ArgSlice[] keys, ref GarnetObjectStoreOutput outputFooter);
Is this okay?

File: libs/server/Resp/Objects/SetCommands.cs (hunk 2 above)
Comment (tisilent) [reply libs/server/Resp/Objects/SetCommands.cs:471]: Parse the keys in Input at the storage layer?

**Comment by tisilent** [comment]
@TalZaccai Hi. I referred to the SortedSetDifference command and I need to reconsider.

File: libs/server/API/IGarnetApi.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (TalZaccai) [reply libs/server/API/IGarnetApi.cs:1195]: Yes that's fine

File: libs/server/Resp/Objects/SetCommands.cs (hunk 2 above)
Comment (TalZaccai) [reply libs/server/Resp/Objects/SetCommands.cs:471]: Continue to parse the keys here, just update the call to SetDiff to output a GarnetObjectStoreOutput

**Comment by TalZaccai** [comment]
@tisilent My apologies, I think I misspoke with some of the previous comments I've had, now that I'm looking at similar command implementations. Let me take another look at this PR...

File: libs/server/Resp/Objects/SetCommands.cs (hunk 4 above)
Comment (tisilent) [reply libs/server/Resp/Objects/SetCommands.cs:520]: Not used. Is it for any other reason.

**Comment by tisilent** [comment]
@TalZaccai Conflict resolved.

File: libs/server/Storage/Session/ObjectStore/SetOps.cs [hunk 9]
```diff
@@ -442,5 +445,152 @@ public GarnetStatus SetIsMember<TObjectContext>(byte[] key, ArgSlice input, ref
...
```
Comment (TalZaccai) [inline libs/server/Storage/Session/ObjectStore/SetOps.cs:555]: Please rename to SetDiff

**Review by TalZaccai — CHANGES_REQUESTED** [review]
Did another pass on this one, let me know if you have any questions :)

File: libs/server/API/GarnetStatus.cs [hunk 10]
```diff
@@ -19,6 +19,10 @@ public enum GarnetStatus : byte
         /// <summary>
         /// Moved
         /// </summary>
-        MOVED
+        MOVED,
+        /// <summary>
+        /// WRONGTYPE
+        /// </summary>
+        WRONGTYPE
```
[the remark this answers is not in the document]
Comment (tisilent) [reply libs/server/API/GarnetStatus.cs]: I removed it.

File: libs/server/Storage/Session/ObjectStore/SetOps.cs [hunk 11]
```diff
@@ -442,5 +445,152 @@ public GarnetStatus SetIsMember<TObjectContext>(byte[] key, ArgSlice input, ref
...
+
+                if (status == GarnetStatus.WRONGTYPE)
+                {
+                    return status;
+                }
+
+                var asMembers = new ArgSlice[diffSet.Count];
+                for (var i = 0; i < diffSet.Count; i++)
+                {
+                    asMembers[i] = scratchBufferManager.CreateArgSlice(diffSet.ElementAt(i));
```
[the remark this answers is not in the document]
Comment (tisilent) [reply libs/server/Storage/Session/ObjectStore/SetOps.cs:525]: Thank you, i was careless.

File: libs/server/Storage/Session/ObjectStore/SetOps.cs [hunk 12]
```diff
@@ -442,5 +445,152 @@ public GarnetStatus SetIsMember<TObjectContext>(byte[] key, ArgSlice input, ref
...
+            for (var i = 1; i < keys.Length; i++)
+            {
+                status = GET(keys[i].ToArray(), out var next, ref objectContext);
+                if (status == GarnetStatus.OK)
+                {
+                    if (next.garnetObject is SetObject nextObject)
+                    {
+                        var nextSet = nextObject.Set;
+                        var interItems = result.Intersect(nextSet, nextSet.Comparer);
+                        result.ExceptWith(interItems);
```
[the remark this answers is not in the document]
Comment (tisilent) [reply libs/server/Storage/Session/ObjectStore/SetOps.cs]: You're correct.

File: libs/server/Storage/Session/ObjectStore/SetOps.cs [hunk 13]
```diff
@@ -552,22 +555,23 @@ public GarnetStatus SetDiffStore(byte[] key, ArgSlice[] keys, out int count)
...
         }
 
-        private (HashSet<byte[]>, GarnetStatus) _setDiff<TObjectContext>(ArgSlice[] keys, ref TObjectContext objectContext)
+        private HashSet<byte[]> SetDiff<TObjectContext>(ArgSlice[] keys, ref TObjectContext objectContext)
             where TObjectContext : ITsavoriteContext<byte[], IGarnetObject, SpanByte, GarnetObjectStoreOutput, long>
         {
-            var result = new HashSet<byte[]>();
+            if (keys.Length == 0)
+            {
+                return [];
```
Comment (TalZaccai) [inline libs/server/Storage/Session/ObjectStore/SetOps.cs:563]: Why not return result here (i.e. default?)

File: libs/server/Storage/Session/ObjectStore/SetOps.cs [hunk 14]
```diff
@@ -579,18 +583,12 @@ public GarnetStatus SetDiffStore(byte[] key, ArgSlice[] keys, out int count)
...
-                    else
-                    {
-                        return (result, GarnetStatus.WRONGTYPE);
+                        result.ExceptWith(nextObject.Set);
                     }
                 }
             }
 
-            return (result, GarnetStatus.OK);
+            return result ?? [];
```
Comment (TalZaccai) [inline libs/server/Storage/Session/ObjectStore/SetOps.cs:591]: Same here, why not just return result?

File: libs/server/Resp/Objects/SetCommands.cs [hunk 15]
```diff
@@ -622,5 +622,112 @@ private unsafe bool SetRandomMember<TGarnetApi>(int count, byte* ptr, ref TGarne
...
+                }
+            }
+
+            // Move input head
+            readHead = (int)(ptr - recvBufferPtr);
+
+            return true;
+        }
+
+        private unsafe bool SetDiffStore<TGarnetApi>(int count, byte* ptr, ref TGarnetApi storageApi)
```
Comment (TalZaccai) [inline libs/server/Resp/Objects/SetCommands.cs:684]: unsafe keyword not necessary here

File: libs/server/Resp/Objects/SetCommands.cs [hunk 16]
```diff
@@ -622,5 +622,112 @@ private unsafe bool SetRandomMember<TGarnetApi>(int count, byte* ptr, ref TGarne
...
+
+        /// <summary>
+        /// Returns the members of the set resulting from the difference between the first set and all the successive sets.
+        /// </summary>
+        /// <typeparam name="TGarnetApi"></typeparam>
+        /// <param name="count"></param>
+        /// <param name="ptr"></param>
+        /// <param name="storageApi"></param>
+        /// <returns></returns>
+        private unsafe bool SetDiff<TGarnetApi>(int count, byte* ptr, ref TGarnetApi storageApi)
```
Comment (TalZaccai) [inline libs/server/Resp/Objects/SetCommands.cs:634]: unsafe keyword not necessary here

File: test/Garnet.test/RespSetTest.cs [hunk 17]
```diff
@@ -273,8 +273,86 @@ public void CanDoSScanWithCursor()
...
+
+            _ = db.SetAdd(key1, key1Value);
+            _ = db.SetAdd(key2, key2Value);
+
+            var result = (RedisResult[])db.Execute("SDIFF", key1, key2);
+            Assert.AreEqual(3, result.Length);
+
+            Assert.IsTrue(Array.Exists(result, t => t.ToString().Equals("a")));
+            Assert.IsTrue(Array.Exists(result, t => t.ToString().Equals("b")));
+            Assert.IsTrue(Array.Exists(result, t => t.ToString().Equals("d")));
```
Comment (TalZaccai) [inline test/Garnet.test/RespSetTest.cs:296]: Might be a bit cleaner:

var strResult = result.Select(r => r.ToString()).ToArray();
var expectedResult = new[] { "a", "b", "d" };
Assert.IsTrue(strResult.OrderBy(t => t).SequenceEqual(expectedResult.OrderBy(t => t)));

File: test/Garnet.test/RespSetTest.cs [hunk 18]
```diff
@@ -273,8 +273,86 @@ public void CanDoSScanWithCursor()
...
+            using var redis = ConnectionMultiplexer.Connect(TestUtils.GetConfig());
+            var db = redis.GetDatabase(0);
+
+            var key1 = "key1";
+            var key1Value = new RedisValue[] { "a", "b", "c", "d" };
+
+            var key2 = "kye2";
+            var key2Value = new RedisValue[] { "c" };
+
+            _ = db.SetAdd(key1, key1Value);
```
Comment (TalZaccai) [inline test/Garnet.test/RespSetTest.cs:288]: Would probably be good to assert that SetAdd was successful (i.e. Assert.AreEqual(1, result));

File: test/Garnet.test/RespSetTest.cs [hunk 19]
```diff
@@ -273,8 +273,86 @@ public void CanDoSScanWithCursor()
...
+        [Test]
+        public void CanDoSdiff()
+        {
+            using var redis = ConnectionMultiplexer.Connect(TestUtils.GetConfig());
+            var db = redis.GetDatabase(0);
+
+            var key1 = "key1";
+            var key1Value = new RedisValue[] { "a", "b", "c", "d" };
+
+            var key2 = "kye2";
```
Comment (TalZaccai) [inline test/Garnet.test/RespSetTest.cs:285]: nit: typo - "key2"

File: test/Garnet.test/RespSetTest.cs [hunk 20]
```diff
@@ -273,8 +273,86 @@ public void CanDoSScanWithCursor()
...
+            _ = db.SetAdd(key2, key2Value);
+
+            var result = (RedisResult[])db.Execute("SDIFF", key1, key2);
+            Assert.AreEqual(3, result.Length);
+
+            Assert.IsTrue(Array.Exists(result, t => t.ToString().Equals("a")));
+            Assert.IsTrue(Array.Exists(result, t => t.ToString().Equals("b")));
+            Assert.IsTrue(Array.Exists(result, t => t.ToString().Equals("d")));
+
+            Assert.IsFalse(Array.Exists(result, t => t.ToString().Equals("c")));
```
Comment (TalZaccai) [inline test/Garnet.test/RespSetTest.cs:298]: Assert.IsFalse(strResult.Contains("c"));

File: test/Garnet.test/RespSetTest.cs [hunk 21]
```diff
@@ -273,8 +273,86 @@ public void CanDoSScanWithCursor()
...
+        {
+            using var redis = ConnectionMultiplexer.Connect(TestUtils.GetConfig());
+            var db = redis.GetDatabase(0);
+
+            var key = "key";
+
+            var key1 = "key1";
+            var key1Value = new RedisValue[] { "a", "b", "c", "d" };
+
+            var key2 = "kye2";
```
Comment (TalZaccai) [inline test/Garnet.test/RespSetTest.cs:326]: nit: typo - "key2"

**Review by TalZaccai — CHANGES_REQUESTED** [review]
Looks good, couple more nits...

File: libs/server/Storage/Session/ObjectStore/SetOps.cs (hunk 13 above)
Comment (tisilent) [reply libs/server/Storage/Session/ObjectStore/SetOps.cs:563]: I wanted to reduce instantiation, but later on, I realized that there were some unavoidable.

**Comment by tisilent** [comment]
@TalZaccai Next time, I will be careful...

**Review by TalZaccai — APPROVED** [review]

**Pull request merged**